ksql-rest-app/src/main/java/io/confluent/ksql/rest/server/StandaloneExecutor.java

```diff
@@ -115,36 +169,86 @@ private void showWelcomeMessage() {
   }
 
   private void executeStatements(final String queries) {
-    final List<QueryMetadata> queryMetadataList = ksqlEngine.createQueries(queries, ksqlConfig);
-    for (QueryMetadata queryMetadata : queryMetadataList) {
-      if (queryMetadata instanceof PersistentQueryMetadata) {
-        PersistentQueryMetadata persistentQueryMetadata = (PersistentQueryMetadata) queryMetadata;
-        persistentQueryMetadata.start();
-      } else {
+    final List<Pair<String, Statement>> statementPairs =
+        ksqlEngine.parseStatements(queries, ksqlEngine.getMetaStore().clone());
+    for (final Pair<String, Statement> statementPair: statementPairs) {
+      final String statementString = statementPair.getLeft();
+      final Statement statement = statementPair.getRight();
+      final BiConsumer<String, Statement> function = functionMap.get(statement.getClass());
+      if (function == null) {
         final String message = String.format(
             "Ignoring statements: %s"
-            + "%nOnly CREATE statements can run in standalone mode.",
-            queryMetadata.getStatementString()
+                + "%nOnly DDL (CREATE STREAM/TABLE, DROP STREAM/TABLE, SET, UNSET) "
+                + "and DML(CSAS, CTAS and INSERT INTO) statements can run in standalone mode.",
+            statementPair.getLeft()
         );
         System.err.println(message);
         log.warn(message);
       }
+      functionMap.get(statement.getClass()).accept(statementPair.getLeft(),
+          statementPair.getRight());
+    }
+
+  }
+
+
+  private void handlePersistentQuery(
+      final Statement statement,
+      final String statementString,
+      final Map<String, Object> configProperties) {
+    final Query query = getQueryFromStatement(statementString, statement);
+    final QueryMetadata queryMetadata =
+        ksqlEngine.getQueryExecutionPlan(query, ksqlConfig);
+    if (statement instanceof CreateAsSelect) {
+      validateCsasCtas(queryMetadata, (CreateAsSelect) statement);
+    }
+    final List<QueryMetadata> queryMetadataList =
+        ksqlEngine.buildMultipleQueries(statementString, ksqlConfig, configProperties);
+    if (queryMetadataList.size() != 1
+        || !(queryMetadataList.get(0) instanceof PersistentQueryMetadata)) {
+      throw new KsqlException("Could not build the query: " + statementString);
+    }
+    queryMetadataList.get(0).start();
+  }
+
+  private Query getQueryFromStatement(final String statementString, final Statement statement) {
+    if (statement instanceof CreateAsSelect) {
+      return  ((CreateAsSelect) statement).getQuery();
+    } else if (statement instanceof InsertInto) {
+      return  ((InsertInto) statement).getQuery();
+    } else {
+      throw new KsqlException("Only CSAS/CTAS and INSERT INTO are persistent queries: "
+          + statementString);
+    }
+  }
+
+  private void validateCsasCtas(final QueryMetadata queryMetadata,
```

Comment:
Shouldn't the engine be doing this validation?

Reply:
Good question!

Reply:
Yes, we can move it to the engine. But after merging #1646 we won't need this validation anymore. So this will be removed anyway.